Mathlib/Algebra/Algebra/Subalgebra/Centralizer.lean

```diff
@@ -0,0 +1,175 @@
+/-
+Copyright (c) 2024 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang
+-/
+import Mathlib.Algebra.Algebra.Subalgebra.Basic
+import Mathlib.RingTheory.TensorProduct.Basic
+import Mathlib.RingTheory.Adjoin.Basic
+import Mathlib.LinearAlgebra.TensorProduct.Basis
+import Mathlib.LinearAlgebra.FreeModule.Basic
+
+/-!
+# Properties of centers and centralizers
+
+This file contains theorems about the center and centralizer of a subalgebra.
+
+## Main results
+
+Let `R` be a commutative ring and `A` and `B` two `R`-algebras.
+- `Subalgebra.centralizer_sup`: if `S` and `T` are subalgebras of `A`, then the centralizer of
+  `S ⊔ T` is the intersection of the centralizer of `S` and the centralizer of `T`.
+- `Subalgebra.centralizer_range_includeLeft_eq_center_tensorProduct`: if `B` is free as a module,
+  then the centralizer of `A` in `A ⊗ B` is `C(A) ⊗ B` where `C(A)` is the center of `A`.
+- `Subalgebra.centralizer_range_includeRight_eq_center_tensorProduct`: if `A` is free as a module,
+  then the centralizer of `B` in `A ⊗ B` is `A ⊗ C(B)` where `C(B)` is the center of `B`.
+-/
+
+namespace Subalgebra
+
+open Algebra.TensorProduct
+
+section CommSemiring
+
+variable {R : Type*} [CommSemiring R]
+variable {A : Type*} [Semiring A] [Algebra R A]
+
+lemma le_centralizer_iff (S T : Subalgebra R A) : S ≤ centralizer R T ↔ T ≤ centralizer R S :=
+  ⟨fun h t ht _ hs ↦ (h hs t ht).symm, fun h s hs _ ht ↦ (h ht s hs).symm⟩
+
+lemma centralizer_sup (S T : Subalgebra R A) :
+    centralizer R ((S ⊔ T : Subalgebra R A) : Set A) = centralizer R S ⊓ centralizer R T :=
+  eq_of_forall_le_iff fun K ↦ by
+    simp_rw [le_centralizer_iff, sup_le_iff, le_inf_iff, K.le_centralizer_iff]
+
+lemma centralizer_iSup {ι : Sort*} (S : ι → Subalgebra R A) :
+    centralizer R ((⨆ i, S i : Subalgebra R A) : Set A) = ⨅ i, centralizer R (S i) :=
+  eq_of_forall_le_iff fun K ↦ by
+    simp_rw [le_centralizer_iff, iSup_le_iff, le_iInf_iff, K.le_centralizer_iff]
+
+end CommSemiring
+
+section Free
+
+variable (R : Type*) [CommSemiring R]
+variable (A : Type*) [Semiring A] [Algebra R A]
+variable (B : Type*) [Semiring B] [Algebra R B]
+
+open Finsupp TensorProduct
+
+/--
+Let `R` be a commutative ring and `A, B` be `R`-algebras where `B` is free as `R`-module.
+For any subalgebra `S` of `A`, the centralizer of `S ⊆ A ⊗ B` is `C_S(A) ⊗ B` where `C_S(A)` is the
+centralizer of `S` in `A`.
+-/
+lemma centralizer_range_includeLeft_comp_val_eq_center_tensorProduct
+    (S : Subalgebra R A) [Module.Free R B] :
+    Subalgebra.centralizer R
+      (Algebra.TensorProduct.includeLeft.comp S.val : S →ₐ[R] A ⊗[R] B).range =
+    (Algebra.TensorProduct.map (Subalgebra.centralizer R (S : Set A)).val
+      (AlgHom.id R B)).range := by
```

Comment:
You can use [Subalgebra.mulMap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/TensorProduct/Subalgebra.html#Subalgebra.mulMap), but maybe ⊤.val is not as desirable as `AlgHom.id R B`.